src/app/core/components/manage-tags/manage-tags.component.spec.ts

```diff
@@ -0,0 +1,159 @@
+  imports: [], // other modules your module needs.
+  providers: [ApiService], // providers available to your module.
+  bootstrap: [] // bootstrap this root component.
+})
+class DummyModule {};
+
+let throwError: boolean = false;
+let deleted: boolean = false;
+let deletedTag: TagVO;
+let renamed: boolean = false;
+let renamedTag: TagVO;
+const mockApiService = {
+  tag: {
+    delete: async (data: any) => {
+      if (throwError) {
+        throw 'Test Error';
+      }
+      deleted = true;
+      deletedTag = data as TagVO;
+      return {
+        getTagVOData: () => {
+          return data;
+        }
+      };
+    },
+    update: async (data: any) => {
+      if (throwError) {
+        throw 'Test Error';
+      }
+      renamed = true;
+      renamedTag = data as TagVO;
+      return {
+        getTagVOData: () => {
+          return data;
+        }
+      };
+    }
+  }
+}
+
+describe('ManageTagsComponent #manage-tags', () => {
+  let shallow: Shallow<ManageTagsComponent>;
+  let defaultTags: TagVO[] = [];
+  async function defaultRender(tags: TagVO[] = defaultTags) {
+    return await shallow.render(`<pr-manage-tags [tags]="tags"></pr-manage-tags>`, {
+      bind: {
+        tags,
+      }
+    });
+  }
+  beforeEach(() => {
+    throwError = false;
+    deleted = false;
+    renamed = false;
+    deletedTag = null;
+    renamedTag = null;
+    defaultTags = [
+      new TagVO({
+        name: 'Tomato',
+        tagId: 2,
+      }),
+      new TagVO({
+        name: 'Potato',
+        tagId: 1,
+      }),
+    ];
+    shallow = new Shallow(ManageTagsComponent, DummyModule).mock(ApiService, mockApiService);
+  });
+
+  it('should exist', async () => {
+    const { element } = await shallow.render();
+    expect(element).not.toBeNull();
+  });
+
+  it('should have a sorted list of tags', async () => {
+    const { find, element } = await defaultRender();
+    expect(find('.tag').length).toBe(2);
+    expect(find('.tag')[0].nativeElement.textContent).toContain('Potato');
+  });
+
+  it('should have a delete button for each tag', async () => {
+    const { find, fixture, element, outputs } = await defaultRender();
+    expect(find('.delete').length).toBeGreaterThan(0);
+    expect(outputs.refreshTags.emit).not.toHaveBeenCalled();
+    find('.delete')[0].nativeElement.click();
+    await fixture.whenStable();
+    expect(deleted).toBeTruthy();
+    expect(deletedTag.name).toBe('Potato');
+    expect(outputs.refreshTags.emit).toHaveBeenCalled();
+    await fixture.detectChanges();
+    expect(find('.tag').length).toBe(1);
+  });
+
+  it('should be able to rename tags', async () => {
+    const { find, fixture, element, outputs } = await defaultRender();
+    expect(find('.edit').length).toBeGreaterThan(0);
+    find('.edit')[0].nativeElement.click();
+    await fixture.detectChanges();
+    expect(find('.tag input').length).toBe(1);
+    expect(find('.tag input').nativeElement.value).toBe('Potato');
```

Comment:
Having these `expect`s in the middle of the test method feels a bit strange to me. The structure I learned for tests was [arrange, act, assert](http://wiki.c2.com/?ArrangeActAssert), and the advantage of following that pattern is that tests become easier to read and easier to think about. Extra asserts in the middle of a method tend to mean that more tests would be better; here, I think we would benefit from, perhaps, "should have an edit button" and "should enter edit mode".